Prokhorova_lab5

```diff
@@ -0,0 +1,724 @@
+}
+
+void FileItem::remove() { }
+
+FileItem::~FileItem() { }
+
+void Directory::addFile(File * file)
+{
+	checkName(file->getName());
+	_files.push_back(file);
+	set_parent(this, file);
+}
+
+void Directory::addDirectory(Directory * directory)
+{
+	checkName(directory->getName());
+	_directories.push_back(directory);
+	directory->_parent = this;
+}
+
+void Directory::addLabel(Label * label)
+{
+	checkName(label->getName());
+	_labels.push_back(label);
+	set_parent(this, label);
+}
+
+Directory* Directory::findDirectory(string name) const
+{
+	for (int i = 0; i < _directories.size(); i++)
+	{
+		if (_directories[i]->_name == name)
+			return _directories[i];
+	}
+	return nullptr;
+}
+
+File* Directory::findFile(string name) const
+{
+	for (int i = 0; i < _files.size(); i++)
+	{
+		if (_files[i]->getName() == name)
+			return _files[i];
+	}
+	return nullptr;
+}
+
+Label * Directory::findLabel(string name) const
+{
+	for (int i = 0; i < _labels.size(); i++)
+	{
+		if (_labels[i]->getName() == name)
+			return _labels[i];
+	}
+	return nullptr;
+}
+
+void Directory::printNames() const
+{
+	std::cout << "Directories:\n";
+	for (int i = 0; i < _directories.size(); i++)
+		cout << _directories[i]->_name << std::endl;
+	std::cout << "Files:\n";
+	for (int i = 0; i < _files.size(); i++)
+		cout << _files[i]->getName() << std::endl;
+	std::cout << "Labels:\n";
+	for (int i = 0; i < _labels.size(); i++)
+		cout << _labels[i]->getName() << std::endl;
+}
+
+void Directory::removeFile(string name)
+{
+	for (int i = 0; i < _files.size(); i++)
+	{
+		if (_files[i]->getName() == name)
+		{
+			_files.erase(_files.begin() + i);
+			return;
+		}
+	}
+	throw runtime_error("File is not found");
+}
+
+void Directory::removeLabel(string name)
+{
+	for (int i = 0; i < _labels.size(); i++)
+	{
+		if (_labels[i]->getName() == name)
+		{
+			_labels.erase(_labels.begin() + i);
+			return;
+		}
+	}
+	throw runtime_error("Label is not found");
+}
+
+void Directory::removeDirectory(string name)
+{
+	for (int i = 0; i < _directories.size(); i++)
+	{
+		if (_directories[i]->_name == name)
+		{
+			_directories.erase(_directories.begin() + i);
+			return;
+		}
+	}
+	throw runtime_error("Directory is not found");
+}
+
+void Directory::remove()
+{
+	_files.clear();
+	_labels.clear();
+	for (int i = 0; i < _directories.size(); i++)
+		_directories[i]->remove();
+	_parent->removeDirectory(_name);
+}
+
+void Directory::checkName(string name) const
+{
+	for (int i = 0; i < _directories.size(); i++)
+	{
+		if (_directories[i]->getName() == name)
+			throw logic_error("Item with this name is already exist");
+	}
+	for (int i = 0; i < _files.size(); i++)
+	{
+		if (_files[i]->getName() == name)
+			throw logic_error("Item with this name is already exist");
+	}
+	for (int i = 0; i < _labels.size(); i++)
+	{
+		if (_labels[i]->getName() == name)
+			throw logic_error("Item with this name is already exist");
+	}
+}
+
+Directory::~Directory()
+{
+	_files.clear();
```

Comment:
Да, точно. Можно сделать таким образом?

Directory::~Directory()
{
for (int i = 0; i < _files.size(); i++)
delete _files[i];
_files.clear();
for (int i = 0; i < _labels.size(); i++)
delete _labels[i];
_labels.clear();
for (int i = 0; i < _directories.size(); i++)
delete _directories[i];
_directories.clear();
}

>Четверг,  5 мая 2016, 20:23 +03:00 от Mironenko <notifications@github.com>:
>
>In  Prokhorova_lab5 :
>> +	}
> +	for (int i = 0; i < _files.size(); i++)
> +	{
> +		if (_files[i]->getName() == name)
> +			throw logic_error("Item with this name is already exist");
> +	}
> +	for (int i = 0; i < _labels.size(); i++)
> +	{
> +		if (_labels[i]->getName() == name)
> +			throw logic_error("Item with this name is already exist");
> +	}
> +}
> +
> +Directory::~Directory()
> +{
> +	_files.clear();
>В векторе указатели. Надо сначала память по ним освободить.
>—
>You are receiving this because you authored the thread.
>Reply to this email directly or  view it on GitHub